### PR TITLE
Add CheckGradient method [checkgradient]

### PR DESCRIPTION
**Pull request by jandrej**
Using forward finite differences on the input vector x, compute the gradient and compare it to the gradient defined in the Operator. The true dofs in ess_tdof are ignored, s.t. gradients with applied essential boundary conditions can be tested. The method will inform the user which column has an l2 norm above the tolerance.

### CODE REVIEW

**Comment by goxberry** [comment]
I'm a bit confused about the terminology MFEM is using. Isn't the gradient the adjoint of the Jacobian (i.e., the Fréchet derivative)?

**Comment by jandrej** [comment]
My naming choice might not be the best here. The NewtonSolver calls the GetGradient method of the attached Operator. And this method checks how this Gradient is comparing against a finite difference version. It is meant as a convenience method to see if the linearization of a nonlinear PDE is implemented correctly.

File: linalg/solvers.cpp [hunk 1]
```diff
@@ -1319,6 +1319,54 @@ void NewtonSolver::Mult(const Vector &b, Vector &x) const
    final_norm = norm;
 }
 
+void NewtonSolver::CheckJacobian(Vector &x, Array<int> &ess_tdof)
+{
+   MFEM_ASSERT(oper != nullptr, "Call SetOperator first.");
+
+   double eps = 1e-8;
```
Comment (barker29) [inline linalg/solvers.cpp]: I recommend making this an argument to the method, so users can decide what tolerance they want to accept. (Apologies for the random drive-by comment from someone who is not reviewing the PR.)

File: linalg/solvers.cpp [hunk 2]
```diff
@@ -1319,6 +1319,54 @@ void NewtonSolver::Mult(const Vector &b, Vector &x) const
...
+      if (ess_tdof.FindSorted(j) >= 0)
+      {
+         jac_col = 0.0;
+      }
+      else
+      {
+         basis(j) = 1.0;
+
+         // Forward finite differences
+         xpert(j) += eps;
```
Comment (goxberry) [inline linalg/solvers.cpp]: It looks like you're forming a finite difference approximation $Df(x)e_{j} = \frac{f(x + h) - f(x)}{\|h\|}$ for some appropriate norm. If I pass in a `GridFunction gf` associated with, for example, an element of $H^{1}$, into `CheckJacobian` as its vector argument `x`, it seems like this code assumes in lines 1349 to 1352 that the norm of `xpert - x` is `eps`. Is that assumption true? It looks like everything is correct if the object passed by reference into `Vector& x` is not associated with a `FiniteElementSpace` -- as I understand it, this would correspond to `x` being an element of $\mathbb{R}^{n}$ for some $n$.

File: linalg/solvers.cpp [hunk 3]
```diff
@@ -1319,6 +1319,54 @@ void NewtonSolver::Mult(const Vector &b, Vector &x) const
    final_norm = norm;
 }
 
+void NewtonSolver::CheckJacobian(Vector &x, Array<int> &ess_tdof)
```
Comment (vladotomov) [inline linalg/solvers.cpp]: Why is this function in NewtonSolver and not in Operator?

Comment (jandrej) [reply linalg/solvers.cpp]: I discussed the same idea with @v-dobrev today and we concluded this method fits better in the Operator itself.

This also enables to implement the action of the gradient using finite differences for tests in the solver method.

File: linalg/solvers.cpp [hunk 4]
```diff
@@ -1319,6 +1319,54 @@ void NewtonSolver::Mult(const Vector &b, Vector &x) const
...
+   xpert = x;
+
+   Operator &jac = oper->GetGradient(x);
+
+   oper->Mult(x, fx);
+
+   for (int j = 0; j < height; j++)
+   {
+      // Is column associated with an essential boundary?
+      if (ess_tdof.FindSorted(j) >= 0)
```
Comment (stefanozampini) [inline linalg/solvers.cpp]: this method is expensive, there's no need to use this optimization and assume ess_tdof is sorted

File: linalg/solvers.cpp [hunk 5]
```diff
@@ -1319,6 +1319,54 @@ void NewtonSolver::Mult(const Vector &b, Vector &x) const
    final_norm = norm;
 }
 
+void NewtonSolver::CheckJacobian(Vector &x, Array<int> &ess_tdof)
+{
+   MFEM_ASSERT(oper != nullptr, "Call SetOperator first.");
+
+   double eps = 1e-8;
+   Vector xpert(height), one(height), fx(height),
```
Comment (stefanozampini) [inline linalg/solvers.cpp]: You should check the sizes here. If the method belongs to the Operator class, then it may be the case that width != height. The Frechet derivative you are testing against is still well defined.

**Comment by tzanio** [comment]
Do you want to keep the branch?
